Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,878 @@
+/-
+Copyright (c) 2024 María Inés de Frutos Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos Fernández, Xavier Généreux
+-/
+import Mathlib.Algebra.Algebra.Equiv
+import Mathlib.Algebra.Algebra.NonUnitalHom
+import Mathlib.Algebra.BigOperators.Finsupp
+import Mathlib.Algebra.Module.BigOperators
+import Mathlib.LinearAlgebra.Finsupp
+
+/-!
+# Skew Monoid algebras
+This file presents a skewed version of `Mathlib.Algebra.MonoidAlgebra.Basic`.
+
+## Definition
+We define `SkewMonoidAlgebra k G := G →₀ k` attached with a skewed convolution product.
+Here, the product of two elements `f g : SkewMonoidAlgebra k G` is the finitely supported
+function whose value at `a` is the sum of `f x * (x • g y)` over all pairs `x, y`
+such that `x * y = a`.
+
+## Implementation notes
+The associativity of the skewed multiplication depends on the `[MulSemiringAction G k]` instance.
+In particular, this means that unlike in `Mathlib.Algebra.MonoidAlgebra.Basic`, `G` will
+need to be a monoid for most of our uses.
+-/
+
+noncomputable section
+
+open BigOperators
+
+open Finset
+
+open Finsupp hiding single mapDomain
+
+universe u₁ u₂ u₃ u₄
+
+variable (k : Type u₁) (G : Type u₂) (H : Type*) {R : Type*}
+
+/-! ### Multiplicative monoids -/
+
+section
+
+variable [AddCommMonoid k]
+
+/-- The skew monoid algebra over a semiring `k` generated by the monoid `G`.
+It is the type of finite formal `k`-linear combinations of terms of `G`,
+endowed with a skewed convolution product.
+-/
+def SkewMonoidAlgebra : Type max u₁ u₂ := G →₀ k
```

Comment:
Note that @adomani is planning a refactor to `MonoidAlgebra` to make the definition
```suggestion
structure SkewMonoidAlgebra : Type max u₁ u₂ where ofFinsupp ::
  toFinsupp : G →₀ k
```
It might be worth getting this new version "right" the first time rather than increasing the number of things @adomani will want to change!

Reply:
Thanks Eric, we are aware of that and Xavier is working on changing our definition.